ArtOfIllusion/src/artofillusion/script/ExecuteScriptWindow.java

```diff
@@ -30,22 +38,97 @@ public class ExecuteScriptWindow extends BFrame
   private LayoutWindow window;
   private RSyntaxTextArea scriptText;
   private BComboBox languageChoice;
-  private String scriptName;
+  public static final String NEW_SCRIPT_NAME = Translate.text ("untitled");
+  private String scriptPath;
 
-  private static File scriptDir;
-  private static String lastScript;
-  private static String lastScriptName = "Untitled";
+  // QUESTION should this be static? It's actually the last directory used for scripts, 
+  // is there a reason it is shared among all editing windows? 
+  private static File scriptDir= new File(ArtOfIllusion.TOOL_SCRIPT_DIRECTORY);
+  private String language;
+  private final BButton save;
+  private final BButton executeSelected;
+  private final BButton executeToCursor;
+  private static final int EDITORS_OFFSET = 32;
+  private static ArrayList <String> openedScripts = new ArrayList<String> ();
 
-  public ExecuteScriptWindow(LayoutWindow win)
+    private javax.swing.filechooser.FileFilter scriptFileFilter;
+    
+    /**
+     * Adds a script path to the recent scripts list. 
+     * This uses a mapping to the current (now) timestamp ; the file paths are in the order of their timestamps, 
+     * so the older is first. The list is truncated to RecentFiles.MAX_RECENT elements. 
+     * @see artofillusion.RecentFiles#MAX_RECENT
+     * @param filePath 
+     */
+    public static void addRecentScript(String filePath)
+    {
+        final Preferences pref = Preferences.userNodeForPackage(ExecuteScriptWindow.class);
+        final String recentFiles[] = pref.get("recentFiles", "").split(File.pathSeparator);
+        java.util.List<String> newRecentFiles = new ArrayList<String>();
+        newRecentFiles.add (filePath);
+        for (String recentFile : recentFiles) 
+                    if (!recentFile.equals(filePath)) // If the current file already has a timestamp it will be updated below
+                        newRecentFiles.add (recentFile);
+        pref.put("recentFiles", String.join (File.pathSeparator, 
+                (String []) newRecentFiles.subList(0, 10).toArray(new String[0])));
+    }
+    
+    public static String [] getRecentScripts()
+    {
+        final Preferences pref = Preferences.userNodeForPackage(ExecuteScriptWindow.class);
+        final java.util.List<String> recentScripts = Arrays.asList (pref.get("recentFiles", "").split(File.pathSeparator));
+        return recentScripts.toArray(new String [0]);
+    } 
+
+  /** This is used to track the "changed" status of the script being edited. */
+  private class ScriptKeyListener implements KeyListener {
+
+      @Override
+      public void keyTyped(KeyEvent e) {
+          save.setEnabled(true);        
+      }
+
+      @Override
+      public void keyPressed(KeyEvent e) {
+      }
+
+      @Override
+      public void keyReleased(KeyEvent e) {
+      }
+        
+    }
```

Comment:
Hmm... I wonder if we might be better off switching to a `TextEditorPane` and listening for a `PropertyChangedEvent`. Perhaps something to consider when refactoring to a unified script editor.

Reply:
Didn't know this class - thanks for the link. I agree, will try this out.